src/qilib/utils/serialization.py

```diff
@@ -18,61 +18,184 @@
+
+def _encode_bytes(data: Any) -> Dict[str, Any]:
+    return {JsonSerializeKey.OBJECT: bytes.__name__, JsonSerializeKey.CONTENT: data.decode('utf-8')}
+
+
+def _decode_bytes(data: Dict[str, Any]) -> bytes:
+    return data[JsonSerializeKey.CONTENT].encode('utf-8')
+
+
+class Serializer:
+    def __init__(self, encoders=None, decoders=None):
+        self.encoder = Encoder()
+        self.decoder = Decoder()
+
+        if encoders is None:
+            encoders = {}
+        self.encoder.encoders = encoders
+        if decoders is None:
+            decoders = {}
+        self.decoder.decoders = decoders
+
+        self.register_encoder(bytes, _encode_bytes)
+        self.register_decoder(bytes.__name__, _decode_bytes)
+        self.register_encoder(np.ndarray, MongoDataSetIO.encode_numpy_array)
+        self.register_decoder(np.array.__name__, MongoDataSetIO.decode_numpy_array)
+
+    def register_encoder(self, type_: type, encode_func: TransformFunction):
+        """ Registers an encoder for a given type
+
+        Args:
+            type_: The type
+            encode_func: The transform function
+
+        """
+
+        self.encoder.encoders[type_] = encode_func
+
+    def register_decoder(self, type_name: str, decode_func: TransformFunction):
+        """ Registers a decoder for a given type
+
+        Args:
+            type_name: The type name
+            decode_func: The transform function
+        """
+
+        self.decoder.decoders[type_name] = decode_func
+
+    def serialize(self, data):
+        """ Serializes a Python object to JSON
+
+        Args:
+            data: Any Python object
+
+        Returns:
+            JSON encoded string
+
+        """
+
+        return self.encoder.encode(self.transform_data(data))
+
+    def unserialize(self, data):
+        """ Unserializes a JSON string to a Python object
+
+        Args:
+            data: The JSON encoded string
+
+        Returns:
+            A Python object decoded from the JSON string
+        """
+
+        return self.decoder.decode(data)
+
+    def transform_data(self, data: Any) -> Any:
+        """ Recursively transfer a Python object and apply transform functions to it
+
+        Args:
+            data: Any Python object
+
+        Returns:
+            The transformed data
+        """
+
+        if isinstance(data, dict):
+            new = {}
+            for key, value in data.items():
+                if isinstance(value, (dict, list, tuple)):
+                    new[key] = self.transform_data(value)
+                else:
+                    new[key] = self._transform(value)
+
+            return new
+
+        elif isinstance(data, (list,)):
+            new = []
+            for item in data:
+                new.append(self.transform_data(self._transform(item)))
+
+            return new
+
+        return self._transform(data)
+
+    def _transform(self, data):
+        type_ = type(data)
+        if type_ in self.encoder.encoders:
```

Comment:
This will not work for subclasses. A beter check is `isinstance`

Reply:
What do you mean?

Reply:
The following fails:
```
import qilib
import numpy as np

class SubClass(np.ndarray):
    pass

a=np.array([1,2])
j=qilib.utils.serialization.serialize(a)
print(j)

a=SubClass([1])
j=qilib.utils.serialization.serialize(a)
print(j)
```

Reply:
I will add a test for this as I don't think we need this feature. How can a parent encoder/decoder know how the encoding/decoding of its subclass looks like?